### PR TITLE
hover for sidebar

### PR DESCRIPTION
**Pull request by opnsenseuser**
sorry it took so long, but it was quite a job for me!

I will customize the tukan, cicada and opnsense themes too and also upload. 
There are some program lines in the main.css of any themes that are no longer needed.
There are also some important fixes in tukan and cicada and also opnsense (font-size,.....)

That's why it's very important to me. 

Thank you very much Franco.

best regards rené

### CODE REVIEW

File: src/opnsense/www/js/opnsense_theme.js [hunk 1]
```diff
@@ -31,6 +31,12 @@
 $(document).ready(function () {
     // traverse loaded css files
     var toggle_sidebar_loaded = false;
+	var nomouse = "mouseenter mouseleave";
+	var layer1_a = $('#navigation > div > nav > #mainmenu > div > a');
+	var layer1_div = $('#navigation > div > nav > #mainmenu > div > div');
+	var layer2_a = $('#navigation > div > nav > #mainmenu > div > div > a');
```
Comment (fabianfrz) [inline src/opnsense/www/js/opnsense_theme.js]: IDs are unique - the front part before can be omitted

File: src/opnsense/www/js/opnsense_theme.js [hunk 2]
```diff
@@ -40,23 +46,43 @@ $(document).ready(function () {
...
         $("#navigation").toggleClass("col-sidebar-left");
         $("main").toggleClass("col-sm-9 col-sm-push-3 col-lg-10 col-lg-push-2 col-lg-12");
         $(".toggle-sidebar > i").toggleClass("fa-chevron-right fa-chevron-left");
         if ($("#navigation").hasClass("col-sidebar-left")) {
             $(".brand-logo").css("display", "none");
             $(".brand-icon").css("display", "inline-block");
+			trigger_sidebar();
             if (store && window.sessionStorage) {
                 sessionStorage.setItem('toggle_sidebar_preset', 1);
+				transition_duration(0);
```
Comment (fabianfrz) [inline src/opnsense/www/js/opnsense_theme.js]: tabs / spaces

File: src/opnsense/www/js/opnsense_theme.js [hunk 3]
```diff
@@ -40,23 +46,43 @@ $(document).ready(function () {
...
             }
         } else {
             $(".brand-icon").css("display", "none");
             $(".brand-logo").css("display", "inline-block");
             $("#navigation.page-side.col-xs-12.col-sm-3.col-lg-2.hidden-xs").css("width", "");
             if (store && window.sessionStorage) {
                 sessionStorage.setItem('toggle_sidebar_preset', 0);
+				mouse_events_off();
+				transition_duration(350);
```
Comment (fabianfrz) [inline src/opnsense/www/js/opnsense_theme.js]: here too

File: src/opnsense/www/js/opnsense_theme.js [hunk 4]
```diff
@@ -65,35 +91,64 @@ $(document).ready(function () {
         /* navigation toggle */
         $(".toggle-sidebar").click(function () {
             opnsense_sidebar_toggle(true);
-            $(this).blur();
+            $(this).blur();
```
Comment (fabianfrz) [inline src/opnsense/www/js/opnsense_theme.js]: no whitespace after the semicolon

File: src/opnsense/www/js/opnsense_theme.js [hunk 5]
```diff
@@ -65,35 +91,64 @@ $(document).ready(function () {
         /* navigation toggle */
         $(".toggle-sidebar").click(function () {
             opnsense_sidebar_toggle(true);
-            $(this).blur();
+            $(this).blur();			
         });
-
-        /* sidebar mouseevents */
-        $("#navigation").hover(function () {
+		/* sidebar mouseenter */
```
Comment (fabianfrz) [inline src/opnsense/www/js/opnsense_theme.js]: again a tab issue

File: src/opnsense/www/js/opnsense_theme.js [hunk 6]
```diff
@@ -65,35 +91,64 @@ $(document).ready(function () {
...
+																}	
+										});
+		/* sidebar mouseleave */								
+		$("#mainmenu").mouseleave(function () {
+			if ($("#navigation").hasClass("col-sidebar-left")) {
+				layer1_a.attr("aria-expanded","false");
+				layer1_a.next("div").removeClass("in");
+															}
+												});
+		/* on resize - toggle sidebar / main navigation */
```
Comment (fabianfrz) [inline src/opnsense/www/js/opnsense_theme.js]: f*** this is messed up…

File: src/opnsense/www/js/opnsense_theme.js [hunk 7]
```diff
@@ -65,35 +91,64 @@ $(document).ready(function () {
...
+					mouseenter: function() { 	$("#navigation.col-sidebar-left").css("width", "380px");
+												$(this).trigger("click");},
+					mouseleave: function() { 	$("#navigation.col-sidebar-left").css("width", "70px");},
+							});
+				
+				layer2_div.on({
+					mouseenter: function() { 	$("#navigation.col-sidebar-left").css("width", "380px");
+												$(this).trigger("click");},
+					mouseleave: function() { 	$("#navigation.col-sidebar-left").css("width", "70px");},
+							});
```
Comment (fabianfrz) [inline src/opnsense/www/js/opnsense_theme.js]: Two issues:
* https://en.wikipedia.org/wiki/Don%27t_repeat_yourself
* this code layout is awful

**Comment by opnsenseuser** [comment]
I am sorry that the source code is so bad. 
I really do not understand why that is and how I can avoid this in the future. 
what would happen if we take this 1:1 to the opnsense source with these errors?

can someone explain to me how I can determine this when editing in notepad ++?

**Comment by fabianfrz** [comment]
I am an Arch Linux user - Notepad++ is Windows only. Somewhere in the editor you should have the following settings:
Tabs: set it to replace it by 4 Spaces
Encoding: UTF-8 (should be correctly detected)
Line Endings: Unix (LF, \n)

Kate as an example:
screenshot_20180908_232126

The issue can be avoided if you would set the default of your editor correctly.
The tab (\t) is a special control character which is replaced by the editor by an indent but our files use 4 spaces. If as you can see, this is machine dependent - so the position of the brackets is completely misplaced on another one or, as you can see, in the diff of GitHub.

What are the consequences?
* the code becomes unmaintainable (bad readable, in case of python also error prone)
* it makes the review hard
* may also make it harder to debug - for example to place breakpoints

**Comment by opnsenseuser** [comment]
ok, thank you for explaining that to me. can you tell me which encoding should I set?

grafik

**Comment by fabianfrz** [comment]
you are already using the correct one

**Comment by opnsenseuser** [comment]
UTF-8 without BOM should be right?
So, the only problem are the tabs? I´m i right ?

**Comment by fabianfrz** [comment]
yes

**Comment by opnsenseuser** [comment]
I should get that right next time! 

And thanks again for the explanation. that is a really great help for me !!

**Comment by opnsenseuser** [comment]
I would like to upload it again to see if it is at least a little better. I must study!! Thank you Fabian

**Pull request closed without merging**